### PR TITLE
[VL] Fix a bug that may break bhj + exchange validation in a corner case

### PR DESCRIPTION
**Pull request by zhztheplayer**
Previous bhj + exchange joint validation may not work because of bugs. The patch is to fix the broken logic.

### CODE REVIEW

File: gluten-core/src/main/scala/io/glutenproject/extension/columnar/TransformHintRule.scala [hunk 1]
```diff
@@ -551,7 +553,9 @@ case class AddTransformHintRule() extends Rule[SparkPlan] {
             maybeExchange match {
               case Some(exchange @ BroadcastExchangeExec(mode, child)) =>
                 TransformHints.tag(bhj, isBhjTransformable.toTransformHint)
-                TransformHints.tagNotTransformable(exchange, isBhjTransformable)
+                if (!isBhjTransformable.isValid) {
+                  TransformHints.tagNotTransformable(exchange, isBhjTransformable)
+                }
```
Comment (zhztheplayer) [inline gluten-core/src/main/scala/io/glutenproject/extension/columnar/TransformHintRule.scala:558]: Once exchange was already tagged as **transformable**, This code may still re-tag it as **non-transformable** if bhj is considered **non-transformable** in validation at the time.

File: gluten-core/src/main/scala/io/glutenproject/extension/columnar/TransformHintRule.scala [hunk 2]
```diff
@@ -313,8 +313,10 @@ case class AddTransformHintRule() extends Rule[SparkPlan] {
 
   /** Inserts a transformable tag on top of those that are not supported. */
   private def addTransformableTags(plan: SparkPlan): SparkPlan = {
-    addTransformableTag(plan)
-    plan.withNewChildren(plan.children.map(addTransformableTags))
+    // Walk the tree with post-order
+    val out = plan.withNewChildren(plan.children.map(addTransformableTags))
+    addTransformableTag(out)
+    out
```
Comment (zhztheplayer) [inline gluten-core/src/main/scala/io/glutenproject/extension/columnar/TransformHintRule.scala:319]: This ensures bhj is validated after the corresponding exchange

**Review by PHILO-HE — APPROVED** [review]
Thanks!

**Pull request merged**